bgpd/bgp_vty.c

```diff
@@ -6115,6 +6115,549 @@ ALIAS_HIDDEN(no_neighbor_addpath_tx_bestpath_per_as,
+		vpn_leak_postchange(BGP_VPN_POLICY_DIR_TOVPN, afi,
+				    bgp_get_default(), bgp);
+	}
+
+	return CMD_SUCCESS;
+}
+
+DEFUN (vpn_policy_no_rd,
+       vpn_policy_no_rd_cmd,
+       "no rd",
+       NO_STR
+       "Specify route distinguisher\n")
+{
+	VTY_DECLVAR_CONTEXT(bgp, bgp);
+	int ret;
+	int doafi[AFI_MAX] = {0};
+	afi_t afi;
+
+	ret = vpn_policy_afis(vty, doafi);
+	if (ret != CMD_SUCCESS)
+		return ret;
+
+
+	for (afi = 0; afi < AFI_MAX; ++afi) {
+		if (!doafi[afi])
+			continue;
+
+		/* pre-change: un-export vpn routes (vpn->vrf routes unaffected)
+		 */
+		vpn_leak_prechange(BGP_VPN_POLICY_DIR_TOVPN, afi,
+				   bgp_get_default(), bgp);
+
+		UNSET_FLAG(bgp->vpn_policy[afi].flags,
+			   BGP_VPN_POLICY_TOVPN_RD_SET);
+
+		/* post-change: re-export vpn routes */
+		vpn_leak_postchange(BGP_VPN_POLICY_DIR_TOVPN, afi,
+				    bgp_get_default(), bgp);
+	}
+
+	return CMD_SUCCESS;
+}
+
+DEFUN (vpn_policy_label,
+       vpn_policy_label_cmd,
+       "label (0-1048575)",
+       "label value for VRF\n"
+       "Label Value <0-1048575>\n")
+{
+	VTY_DECLVAR_CONTEXT(bgp, bgp);
+	mpls_label_t label;
+	int doafi[AFI_MAX] = {0};
+	afi_t afi;
+	int ret;
+
+	label = strtoul(argv[1]->arg, NULL, 10);
+
+	ret = vpn_policy_afis(vty, doafi);
+	if (ret != CMD_SUCCESS)
+		return ret;
+
+	for (afi = 0; afi < AFI_MAX; ++afi) {
+		if (!doafi[afi])
+			continue;
+
+		/* pre-change: un-export vpn routes (vpn->vrf routes unaffected)
+		 */
+		vpn_leak_prechange(BGP_VPN_POLICY_DIR_TOVPN, afi,
+				   bgp_get_default(), bgp);
+
+		bgp->vpn_policy[afi].tovpn_label = label;
+
+		/* post-change: re-export vpn routes */
+		vpn_leak_postchange(BGP_VPN_POLICY_DIR_TOVPN, afi,
+				    bgp_get_default(), bgp);
+	}
+
+	return CMD_SUCCESS;
+}
+
+DEFUN (vpn_policy_no_label,
+       vpn_policy_no_label_cmd,
+       "no label",
+       "Negate a command or set its defaults\n"
+       "label value for VRF\n")
+{
+	VTY_DECLVAR_CONTEXT(bgp, bgp);
+	int doafi[AFI_MAX] = {0};
+	afi_t afi;
+	int ret;
+
+	ret = vpn_policy_afis(vty, doafi);
+	if (ret != CMD_SUCCESS)
+		return ret;
+
+	for (afi = 0; afi < AFI_MAX; ++afi) {
+		if (!doafi[afi])
+			continue;
+
+		/* pre-change: un-export vpn routes (vpn->vrf routes unaffected)
+		 */
+		vpn_leak_prechange(BGP_VPN_POLICY_DIR_TOVPN, afi,
+				   bgp_get_default(), bgp);
+
+		bgp->vpn_policy[afi].tovpn_label = MPLS_LABEL_NONE;
+
+		/* post-change: re-export vpn routes */
+		vpn_leak_postchange(BGP_VPN_POLICY_DIR_TOVPN, afi,
+				    bgp_get_default(), bgp);
+	}
+
+	return CMD_SUCCESS;
+}
+
+DEFUN (vpn_policy_nexthop,
+       vpn_policy_nexthop_cmd,
+       "nexthop <A.B.C.D|X:X::X:X|self>",
+       "Specify next hop to use for VRF advertised prefixes\n"
+       "IPv4 prefix\n"
+       "IPv6 prefix\n"
+       "Use configured router-id (default)\n")
+{
+	VTY_DECLVAR_CONTEXT(bgp, bgp);
+	int doafi[AFI_MAX] = {0};
+	afi_t afi;
+	int ret;
+	struct prefix p;
+	int nhself = 0;
+
+	if ((argv[1]->arg[0] == 's') || (!str2prefix(argv[1]->arg, &p)))
```

Comment:
argv_find

Reply:
My assumption is that I won't need to go back and hit each place where this is done. @paulzlabn you should be able to find them